Sprint 08 (tableView, tableViewCell, UITableViewDataSource, UITableViewDelegate, Fork)

Dassam opened the pull request:
#8

1. **Created a new project in which**

   - iOS 13.0 is listed as the minimum supported version;
   - in the settings of the main target, assembly on the iPad is disabled;
   - only the portrait orientation support checkbox is checked (in the Device Orientation section), and the corresponding values for Supported interface orientations (iPhone) are removed from Info;
   - the structure of the files is the same as the structure that we considered in the topic "Initialization of the second project";
   - Git is connected for the project - in the Source Control menu in Xcode, you will have access to the repository management menus.


2. **Designed Launch Screen app**
   - the screen must fully match [layout](https://www.figma.com/file/HyDfKh5UVPohPZIhBqIm3q/Image-Feed-(YP)
  
3. **Designed the main screen with a ribbon of images with the following parameters:**
   - the tape contains 20 photos;
   - there is an indent between the photos;
   - the edges of the photos are rounded;
   - a label with the current date has been added to the photo;
   - Added a like button to the photo;
   - like is put down for cells with an even index and not put down - for odd ones;
   - sizes, parameters and indents of elements correspond to layouts;
   - cell height depends on photo size and is dynamic*.
   -
4. **App launches.**
 
5. **UI elements adapt to iPhone screen resolutions**

Dassam committed:
#8
- Added UIColor+Extensions file
- Realised launch screen design

Dassam committed:
#8
- Added fonts
- Added rules for status bar in info post

Dassam committed:
#8

- Made the screen according to the layout
- Implemented the logic needed to display 20 images along with likes
- Implemented Gradient on date
- Added the necessary icons

Dassam committed:
#8

- Made the screen according to the layout
- Implemented the logic needed to display 20 images along with likes
- Implemented Gradient on date
- Added the necessary icons

Dassam committed:
Merge branch 'sprint_08' of https://github.com/Dassam/ImageFeedList into sprint_08

Dassam committed:
#8
- Made the screen according to the layout
- Implemented the logic needed to display 20 images along with likes
- Implemented Gradient on date
- Added the necessary icons

Nikitaaeee commented on ImageFeedList/Presentation/ImagesList/ImagesListViewController.swift:
```diff
@@ -0,0 +1,84 @@
+import UIKit
+
+class ImagesListViewController: UIViewController {
+    @IBOutlet private var tableView: UITableView!
```
Для улучшения производительности можно добавлять модификатор `final` к классам, от которых мы не будем наследоваться. (Как ты сделал в **ImagesListCell**, например) 
В таком случае комплятор будет быстрее находить адрес ячейки памяти при вызове функций.

Если интересно, тут можно подробнее почитать про то, что такое диспетчеризация и как она работает (довольно частая тема на собеседованиях): 
https://habr.com/ru/companies/simbirsoft/articles/673636/

Dassam replied:
Спасибо!

Nikitaaeee commented on ImageFeedList/Presentation/ImagesList/ImagesListViewController.swift:
```diff
@@ -0,0 +1,84 @@
+      let imageViewHeight = imageHeight + imageInsets.top + imageInsets.bottom
+      return imageViewHeight
+    }
+    
```
Выносить методы, реализующие функционал протоколов в отдельный `extension` - отличная идея!
Можно сделать еще красивее:
1. разделить реализацию _**UITableViewDataSource**_ и _**UITableViewDelegate**_ на два разных `extension`. 
2. добавить комментарии к каждому extension такого типа: `//MARK: - UITableViewDataSource`
Добавленные в таком виде комментарии упрощают организацию на Minimap (открывается через Shift+Control+Command+M) и облегчает поиск по коду.

В целом про организацию кода и использование таких комментариев можно дополнительно посмотреть тут:
https://github.com/RedMadRobot/RMR-swift-style-guide#code-organization

Dassam replied:
Разбил, спасибо!

Nikitaaeee commented on ImageFeedList/Presentation/ImagesList/ImagesListCell.swift:
```diff
@@ -0,0 +1,29 @@
+    
+    static let reuseIdentifier = "ImagesListCell"
+    
+    let gradient = CAGradientLayer()
+    let gradientColors = [
+        UIColor(red: 0.102,
+                green: 0.106,
+                blue: 0.133,
+                alpha: 0.2).cgColor,
+        UIColor(red: 0.102,
+                green: 0.106,
+                blue: 0.133,
+                alpha: 0).cgColor
+      ]
+}
```
Доп задание (со звездочкой) с градиентом на ячейке выполнено! 🔥

Dassam replied:
🔥😁

Dassam committed:
#8
- Added the final modifier to the ImagesListViewController class
- Split the implementation of UITableViewDataSource and UITableViewDelegate into two different extensions
- Fixed the table top constraint now it stretches to the top of the screen
- Likes are reversed
- Removed the Selection cell setting - None
- Added settings to Info plst so that the status bar is white
